pcap-ioplugin.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2017
+ *	Internet Systems Consortium, Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that: (1) source code distributions
+ * retain the above copyright notice and this paragraph in its entirety, (2)
+ * distributions including binary code include the above copyright notice and
+ * this paragraph in its entirety in the documentation or other materials
+ * provided with the distribution, and (3) all advertising materials mentioning
+ * features or use of this software display the following acknowledgement:
+ * ``This product includes software developed by the University of California,
+ * Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
+ * the University nor the names of its contributors may be used to endorse
+ * or promote products derived from this software without specific prior
+ * written permission.
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * pcap-ioplugin.c - supports dynamic loading of compression modules
+ *	Created by Ray Bellis, ISC.
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "ftmacros.h"
+
+#ifdef _WIN32
+#include <pcap-stdinc.h>
+#else /* _WIN32 */
+#if HAVE_INTTYPES_H
+#include <inttypes.h>
+#elif HAVE_STDINT_H
+#include <stdint.h>
+#endif
+#ifdef HAVE_SYS_BITYPES_H
+#include <sys/bitypes.h>
+#endif
+#include <sys/types.h>
+#endif /* _WIN32 */
+
+#include <errno.h>
+#include <memory.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "pcap-int.h"
+
+#ifdef HAVE_DLFCN_H
+#include <dlfcn.h>
+#endif
+
+#ifdef HAVE_OS_PROTO_H
+#include "os-proto.h"
+#endif
+
+static FILE*
+stdio_open_read(const char *fname, char *errbuf)
+{
+	FILE *fp = NULL;
+
+	if (fname == NULL) {
+		snprintf(errbuf, PCAP_ERRBUF_SIZE,
+		    "A null pointer was supplied as the file name");
+		return (NULL);
+	}
+
+        if (fname[0] == '-' && fname[1] == '\0')
+        {
+                fp = stdin;
+                if (stdin == NULL) {
+                        snprintf(errbuf, PCAP_ERRBUF_SIZE,
+                                 "The standard input is not open");
```

Comment:
This is not a good test for a safe use of `stdin` since it could be a fixed address (like on djgpp).
Perhaps `if (!stdin || ferror(stdin))` is better.

Reply:
Is djgpp still a build target that matters?  (I'm seriously asking: I didn't think so).
I'm happy to do something different here, but I want to make sure that it's really better.
I don't understand what !stdin would do?  That's the same as stdin != NULL, isn't it?

Reply:
changed based upon suggestion.

Reply:
> Is djgpp still a build target that matters?

Not really. But there could be other exotic targets were `#define stdin (&__internal_stdin)`.

Reply:
On my Solaris 11 VM:

/usr/include/stdio.h includes

    #include <iso/stdio_iso.h>

and /usr/include/iso/stdio_ios.h includes

```
#if defined(__STDC__)
extern __FILE	__iob[_NFILE];
#define	stdin	(&__iob[0])
#define	stdout	(&__iob[1])
#define	stderr	(&__iob[2])
#else
extern __FILE	_iob[_NFILE];
#define	stdin	(&_iob[0])
#define	stdout	(&_iob[1])
#define	stderr	(&_iob[2])
#endif	/* __STDC__ */
```

That goes all the way back to System N, for some Roman-numeral version of N, so other commercial UN\*Xes may do that, too.

(GNU libc doesn't, so few Linux distributions do that; as far as I know, Chris Torek's stdio code doesn't, either, so I don't think any of the BSD-favored UN\*Xes, including macOS, do that.)

Reply:
> I don't understand what !stdin would do? That's the same as stdin != NULL, isn't it?

Yes.